eyelid-match-ops/src/primitives/yashe.rs

```diff
@@ -104,14 +130,93 @@ where
         (priv_key, pub_key)
     }
 
+    /// Encrypt a message m encoded in the polynomial ring
+    pub fn encrypt(
+        &self,
+        mut m: Message<C>,
+        public_key: PublicKey<C>,
+        rng: &mut ThreadRng,
+    ) -> Ciphertext<C> {
+        // TODO: document the equations that are being implemented by each block.
+
+        // Create the ciphertext by sampling error polynomials and applying them to the public key.
+        let s = self.sample_err(rng);
+        let e = self.sample_err(rng);
+        let mut c = s * public_key.h + e;
+
+        // Divide the polynomial coefficient modulus by T, using primitive integer arithmetic.
+        let qdt = C::modulus_as_u128() / C::t_as_u128();
+        let qdt = C::Coeff::from(qdt);
+
+        // Multiply the message by the qdt scalar, and add it to the ciphertext.
+        m.m *= qdt;
+        c += m.m;
+
+        Ciphertext { c }
+    }
+
+    /// Decrypt a ciphertext
+    pub fn decrypt(&self, c: Ciphertext<C>, private_key: PrivateKey<C>) -> Message<C> {
+        // TODO: document the equations that are being implemented by each block.
+
+        // Multiply the ciphertext by the private key polynomial.
+        let mut res = c.c * private_key.priv_key;
+
+        // TODO: is this the equation that is being implemented here?
+        // In primitive integer arithmetic, calculate:
+        // (res[i] * T + (Q - 1)/2) / Q % T
+
+        // Since this equation always results in zero for a zero coefficient, we don't need to
+        // calculate leading zero terms.
+        //
+        // TODO:
+        // consider creating Poly methods which take a closure to update each coefficient
+        // - for leading zero coefficients to MAX_POLY_DEGREE, and only non-zero coeffs_mut()
+        // - for Coeff and u128 arithmetic
+        for coeff in res.coeffs_mut() {
+            // Convert coefficient to a primitive integer
+            let mut coeff_res = C::coeff_as_u128(*coeff);
+
+            // Multiply by T
+            coeff_res *= C::t_as_u128();
+            // Add (Q - 1)/2 to implement rounding rather than truncation
+            coeff_res += C::modulus_minus_one_div_two_as_u128();
+            // Divide by Q
+            coeff_res /= C::modulus_as_u128();
+            // Modulo T
+            coeff_res %= C::t_as_u128();
+
+            // And update the coefficient
+            *coeff = coeff_res.into();
+        }
+
+        // Raw coefficient access must be followed by a truncation check.
+        res.truncate_to_canonical_form();
+
+        Message { m: res }
+    }
+
+    /// Sample a polynomial with small random coefficients using a gaussian distribution.
+    pub fn sample_err(&self, rng: &mut ThreadRng) -> Poly<C> {
+        self.sample_gaussian(C::ERROR_DELTA, rng)
+    }
+
+    /// Sample a polynomial with small random coefficients using a gaussian distribution.
+    /// TODO: this function seems to be returning too few non-zero elements
```

Comment:
`v as i64` will truncate the `f64` value, discarding any fractional parts.

Did you want to [round halfway](https://doc.rust-lang.org/std/primitive.f64.html#method.round) or [use the next highest integer](https://doc.rust-lang.org/std/primitive.f64.html#method.ceil) before converting?

Reply:
round halfway

Reply:
Ok, here’s how to do that:
https://github.com/Inversed-Tech/eyelid/pull/82#discussion_r1602322652

(There’s also round ties to even, but ties are low probability, so the difference doesn’t matter.)